# Getting started with pandas - Intro

This chapter introduces you to the pandas package. It also serves as a test for your technical setup.

**Important Note on pandas 3.0:** This course uses pandas 3.0, which introduces significant changes:

1. **String dtype**: String columns are now inferred as `str` dtype instead of `object` dtype. When you see `dtype: str` in output, this is the new dedicated string type (backed by PyArrow if available). When selecting string columns with `select_dtypes()`, use `select_dtypes(["string"])` or `select_dtypes(["object", "string"])` for compatibility with both pandas 2.x and 3.x.

2. **Copy-on-Write (CoW)**: Now the default behavior. Any subset or derived DataFrame/Series behaves as a copy - changes won't affect the original. This means:
   - Chained assignment (e.g., `df[df['x'] > 5]['y'] = 10`) no longer works (never was a good idea).
   - Always use direct indexing: `df.loc[df['x'] > 5, 'y'] = 10`
   - The `SettingWithCopyWarning` has been removed
   - Defensive `.copy()` calls are less necessary but still useful for clarity

# Load required packages and setup display of dataframes

In [1]:
import pandas as pd

# Loading and inspecting data

## Loading and inspecting a CSV file

In [2]:
df = pd.read_csv("../../data/raw/financial_data_intro.csv")
df.head()

u_company_name_id  u_year                u_company_name  cb_naics u_iso3  \
0              14651    2005  British American Tobacco PLC    312230    GBR   
1              14651    2006  British American Tobacco PLC    312230    GBR   
2              14651    2007  British American Tobacco PLC    312230    GBR   
3              14651    2008  British American Tobacco PLC    312230    GBR   
4              14651    2009  British American Tobacco PLC    312230    GBR   

        u_fye   cb_cusip      cb_at     cb_ni  cb_financial_industry  
0  2005-12-31  110448107  32737.984  2707.110                  False  
1  2006-12-31  110448107  34816.074  3713.506                  False  
2  2007-12-31  110448107  37161.970  4226.559                  False  
3  2008-12-31  110448107  40276.807  3591.888                  False  
4  2009-12-31  110448107  43026.854  4386.107                  False

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 824 entries, 0 to 823
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   u_company_name_id      824 non-null    int64  
 1   u_year                 824 non-null    int64  
 2   u_company_name         824 non-null    str    
 3   cb_naics               824 non-null    int64  
 4   u_iso3                 824 non-null    str    
 5   u_fye                  824 non-null    str    
 6   cb_cusip               824 non-null    str    
 7   cb_at                  824 non-null    float64
 8   cb_ni                  824 non-null    float64
 9   cb_financial_industry  824 non-null    bool   
dtypes: bool(1), float64(2), int64(3), str(4)
memory usage: 89.5 KB


In [4]:
df.shape

(824, 10)

In [5]:
df.columns

Index(['u_company_name_id', 'u_year', 'u_company_name', 'cb_naics', 'u_iso3',
       'u_fye', 'cb_cusip', 'cb_at', 'cb_ni', 'cb_financial_industry'],
      dtype='str')

In [6]:
df.describe()

u_company_name_id       u_year       cb_naics         cb_at  \
count         824.000000   824.000000     824.000000  8.240000e+02   
mean        45233.387136  2012.111650  379140.905340  1.136859e+05   
std         32869.292595     4.166015  207892.935774  2.963489e+05   
min          2172.000000  2005.000000     325.000000  0.000000e+00   
25%         14651.000000  2009.000000  324110.000000  7.634920e+02   
50%         34617.000000  2012.000000  336111.000000  1.647187e+04   
75%         77954.000000  2016.000000  523930.000000  8.486373e+04   
max        109031.000000  2019.000000  999977.000000  2.261780e+06   

              cb_ni  
count    824.000000  
mean    2191.241188  
std     5047.784304  
min   -50119.000000  
25%       -0.312750  
50%      484.862000  
75%     3470.270750  
max    50778.396000

In [7]:
df.describe().transpose()

count           mean            std      min           25%  \
u_company_name_id  824.0   45233.387136   32869.292595   2172.0   14651.00000   
u_year             824.0    2012.111650       4.166015   2005.0    2009.00000   
cb_naics           824.0  379140.905340  207892.935774    325.0  324110.00000   
cb_at              824.0  113685.916945  296348.893884      0.0     763.49200   
cb_ni              824.0    2191.241188    5047.784304 -50119.0      -0.31275   

                           50%           75%          max  
u_company_name_id   34617.0000   77954.00000   109031.000  
u_year               2012.0000    2016.00000     2019.000  
cb_naics           336111.0000  523930.00000   999977.000  
cb_at               16471.8695   84863.72725  2261780.000  
cb_ni                 484.8620    3470.27075    50778.396

## Checking the file specifications

It is important to correctly adapt all settings for every single file. By "chance", the defaults of `pd.read_csv()` were correct for our example above. Generally, it is a good idea to have a look at the first few lines of a text file, especially when the defaults did not produce the expected columns and data types.

> **Note:** Use the [Checklist for data loading issues](../../checklist-data-loading.md)

In [8]:
# Printing the first few lines of the file:
with open("../../data/raw/financial_data_intro.csv") as f:
    for _ in range(5):
        print(f.readline(), end="")

u_company_name_id,u_year,u_company_name,cb_naics,u_iso3,u_fye,cb_cusip,cb_at,cb_ni,cb_financial_industry
14651,2005,British American Tobacco PLC,312230,GBR,2005-12-31,110448107,32737.984,2707.11,FALSE
14651,2006,British American Tobacco PLC,312230,GBR,2006-12-31,110448107,34816.074,3713.506,FALSE
14651,2007,British American Tobacco PLC,312230,GBR,2007-12-31,110448107,37161.97,4226.559,FALSE
14651,2008,British American Tobacco PLC,312230,GBR,2008-12-31,110448107,40276.807,3591.888,FALSE


In [9]:
# Pretend we use the wrong separator (";", which would often be correct for non-english CSV files):
df_wrong = pd.read_csv("../../data/raw/financial_data_intro.csv", sep=";")
df_wrong.head()

u_company_name_id,u_year,u_company_name,cb_naics,u_iso3,u_fye,cb_cusip,cb_at,cb_ni,cb_financial_industry
0  14651,2005,British American Tobacco PLC,312230...                                                      
1  14651,2006,British American Tobacco PLC,312230...                                                      
2  14651,2007,British American Tobacco PLC,312230...                                                      
3  14651,2008,British American Tobacco PLC,312230...                                                      
4  14651,2009,British American Tobacco PLC,312230...

In [10]:
df_wrong.info()

<class 'pandas.DataFrame'>
RangeIndex: 824 entries, 0 to 823
Data columns (total 1 columns):
 #   Column                                                                                                    Non-Null Count  Dtype
---  ------                                                                                                    --------------  -----
 0   u_company_name_id,u_year,u_company_name,cb_naics,u_iso3,u_fye,cb_cusip,cb_at,cb_ni,cb_financial_industry  824 non-null    str  
dtypes: str(1)
memory usage: 72.1 KB


In [11]:
# Pretend we assume no header
df_wrong = pd.read_csv("../../data/raw/financial_data_intro.csv", header=None)
df_wrong.head()

0       1                             2         3       4  \
0  u_company_name_id  u_year                u_company_name  cb_naics  u_iso3   
1              14651    2005  British American Tobacco PLC    312230     GBR   
2              14651    2006  British American Tobacco PLC    312230     GBR   
3              14651    2007  British American Tobacco PLC    312230     GBR   
4              14651    2008  British American Tobacco PLC    312230     GBR   

            5          6          7         8                      9  
0       u_fye   cb_cusip      cb_at     cb_ni  cb_financial_industry  
1  2005-12-31  110448107  32737.984   2707.11                  FALSE  
2  2006-12-31  110448107  34816.074  3713.506                  FALSE  
3  2007-12-31  110448107   37161.97  4226.559                  FALSE  
4  2008-12-31  110448107  40276.807  3591.888                  FALSE

In [12]:
df_wrong.info()

<class 'pandas.DataFrame'>
RangeIndex: 825 entries, 0 to 824
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   0       825 non-null    str  
 1   1       825 non-null    str  
 2   2       825 non-null    str  
 3   3       825 non-null    str  
 4   4       825 non-null    str  
 5   5       825 non-null    str  
 6   6       825 non-null    str  
 7   7       825 non-null    str  
 8   8       825 non-null    str  
 9   9       825 non-null    str  
dtypes: str(10)
memory usage: 123.0 KB


In [13]:
# Clean up
del df_wrong

# Select DataFrame columns

## Selecting a single column

using `[]` with a string returns one column as a `pd.Series`

>**Note**
You can also select a column using the dot notation, e.g. `df.u_company_name`, but this only works if the column name is a valid Python identifier and is not advised for general use.

In [14]:
# select a Series (single column) of a DataFrame
df["u_company_name"]

0      British American Tobacco PLC
1      British American Tobacco PLC
2      British American Tobacco PLC
3      British American Tobacco PLC
4      British American Tobacco PLC
                   ...             
819             Midatech Pharma PLC
820             Midatech Pharma PLC
821                   Motif Bio plc
822                   Motif Bio plc
823                   Motif Bio plc
Name: u_company_name, Length: 824, dtype: str

In [15]:
# Note that the result is a Pandas-Series, not a DataFrame!
type(df["u_company_name"])

pandas.Series

## Selecting multiple columns

using `[]` with a list of strings returns multiple columns as a `pd.DataFrame`.

In [16]:
# Multiple columns are selected by supplying a list:
df[["u_company_name", "u_year"]]

u_company_name  u_year
0    British American Tobacco PLC    2005
1    British American Tobacco PLC    2006
2    British American Tobacco PLC    2007
3    British American Tobacco PLC    2008
4    British American Tobacco PLC    2009
..                            ...     ...
819           Midatech Pharma PLC    2017
820           Midatech Pharma PLC    2018
821                 Motif Bio plc    2016
822                 Motif Bio plc    2017
823                 Motif Bio plc    2018

[824 rows x 2 columns]

In [17]:
# The result is a DataFrame!
type(df[["u_company_name", "u_year"]])

pandas.DataFrame

## Selecting specific types of columns

The `select_dtypes()` method of `pd.DataFrame` allows you to subset the DataFrame by selecting only specific types of columns.


In [18]:
df.select_dtypes(["int", "float"])

u_company_name_id  u_year  cb_naics      cb_at     cb_ni
0                14651    2005    312230  32737.984  2707.110
1                14651    2006    312230  34816.074  3713.506
2                14651    2007    312230  37161.970  4226.559
3                14651    2008    312230  40276.807  3591.888
4                14651    2009    312230  43026.854  4386.107
..                 ...     ...       ...        ...       ...
819              62880    2017    325414     66.595   -21.733
820              62880    2018    325414     26.056   -19.156
821              64528    2016    325414     28.426   -40.324
822              64528    2017    325414     29.188   -44.810
823              64528    2018    325414     18.724   -13.985

[824 rows x 5 columns]

In [19]:
# you can also produce a list of the columns:
numeric_columns = df.select_dtypes(["int", "float"]).columns.to_list()
numeric_columns

['u_company_name_id', 'u_year', 'cb_naics', 'cb_at', 'cb_ni']


## Rule-based column selection

Especially for datasets with many columns, we often need to select columns based on specific patterns or rules.

This can be achieved using various methods, including loops, list comprehensions, and pandas string methods.

Most of the time, we refer to the `df.columns` attribute (a special kind of `pd.Series`)to get the column names to select from.

In [20]:
# df.columns is a special kind of `pd.Series` (an `Index`)
df.columns

Index(['u_company_name_id', 'u_year', 'u_company_name', 'cb_naics', 'u_iso3',
       'u_fye', 'cb_cusip', 'cb_at', 'cb_ni', 'cb_financial_industry'],
      dtype='str')

In [21]:
# lists created by code are especially useful for column selection

# Example: selecting columns starting with "cb_"
selected_cols = []
for c in df.columns:
    if c.startswith("cb_"):
        selected_cols.append(c)
print(f"selected_cols: {selected_cols}")

df[selected_cols]

selected_cols: ['cb_naics', 'cb_cusip', 'cb_at', 'cb_ni', 'cb_financial_industry']


cb_naics   cb_cusip      cb_at     cb_ni  cb_financial_industry
0      312230  110448107  32737.984  2707.110                  False
1      312230  110448107  34816.074  3713.506                  False
2      312230  110448107  37161.970  4226.559                  False
3      312230  110448107  40276.807  3591.888                  False
4      312230  110448107  43026.854  4386.107                  False
..        ...        ...        ...       ...                    ...
819    325414  59564R302     66.595   -21.733                  False
820    325414  59564R302     26.056   -19.156                  False
821    325414  619784101     28.426   -40.324                  False
822    325414  619784101     29.188   -44.810                  False
823    325414  619784101     18.724   -13.985                  False

[824 rows x 5 columns]

In [22]:
# Alternative solutions for generating the same `selected_cols`

# List comprehension:
selected_cols = [c for c in df.columns if c.startswith("cb_")]
print(f"From list comprehension: {selected_cols}")

# Using methods included in pandas:
selected_cols = df.columns[df.columns.str.startswith("cb_")]
print(f"Using pandas methods: {selected_cols}")

From list comprehension: ['cb_naics', 'cb_cusip', 'cb_at', 'cb_ni', 'cb_financial_industry']
Using pandas methods: Index(['cb_naics', 'cb_cusip', 'cb_at', 'cb_ni', 'cb_financial_industry'], dtype='str')


# Select rows from a DataFrame

## Selecting the first or last rows

In [23]:
df.head(10)

u_company_name_id  u_year                u_company_name  cb_naics u_iso3  \
0              14651    2005  British American Tobacco PLC    312230    GBR   
1              14651    2006  British American Tobacco PLC    312230    GBR   
2              14651    2007  British American Tobacco PLC    312230    GBR   
3              14651    2008  British American Tobacco PLC    312230    GBR   
4              14651    2009  British American Tobacco PLC    312230    GBR   
5              14651    2010  British American Tobacco PLC    312230    GBR   
6              14651    2011  British American Tobacco PLC    312230    GBR   
7              14651    2012  British American Tobacco PLC    312230    GBR   
8              14651    2013  British American Tobacco PLC    312230    GBR   
9              14651    2014  British American Tobacco PLC    312230    GBR   

        u_fye   cb_cusip      cb_at     cb_ni  cb_financial_industry  
0  2005-12-31  110448107  32737.984  2707.110                  False  
1  2006-12-31  110448107  34816.074  3713.506                  False  
2  2007-12-31  110448107  37161.970  4226.559                  False  
3  2008-12-31  110448107  40276.807  3591.888                  False  
4  2009-12-31  110448107  43026.854  4386.107                  False  
5  2010-12-31  110448107  42882.112  4431.357                  False  
6  2011-12-31  110448107  42134.790  4808.702                  False  
7  2012-12-31  110448107  44439.167  6246.234                  False  
8  2013-12-31  110448107  44552.569  6470.490                  False  
9  2014-12-31  110448107  40762.953  4852.547                  False

In [24]:
df.tail(10)

u_company_name_id  u_year         u_company_name  cb_naics u_iso3  \
814              24931    2016  DBV Technologies S.A.    325414    FRA   
815              24931    2017  DBV Technologies S.A.    325414    FRA   
816              24931    2018  DBV Technologies S.A.    325414    FRA   
817              62880    2015    Midatech Pharma PLC    325414    GBR   
818              62880    2016    Midatech Pharma PLC    325414    GBR   
819              62880    2017    Midatech Pharma PLC    325414    GBR   
820              62880    2018    Midatech Pharma PLC    325414    GBR   
821              64528    2016          Motif Bio plc    325414    GBR   
822              64528    2017          Motif Bio plc    325414    GBR   
823              64528    2018          Motif Bio plc    325414    GBR   

          u_fye   cb_cusip    cb_at    cb_ni  cb_financial_industry  
814  2016-12-31  23306J101  303.370 -120.853                  False  
815  2017-12-31  23306J101  213.760 -177.557                  False  
816  2018-12-31  23306J101  196.635 -190.140                  False  
817  2015-12-31  59564R302   94.435  -14.892                  False  
818  2016-12-31  59564R302   69.937  -24.874                  False  
819  2017-12-31  59564R302   66.595  -21.733                  False  
820  2018-12-31  59564R302   26.056  -19.156                  False  
821  2016-12-31  619784101   28.426  -40.324                  False  
822  2017-12-31  619784101   29.188  -44.810                  False  
823  2018-12-31  619784101   18.724  -13.985                  False


## Rule-based selection

Passing a `list` or `pd.Series` of **booleans** to `[]` will also select rows (not columns) when used with `[]`

>**Note**: The number of booleans passed to `[]` must equal the number of rows!
>
>e.g. `df[[True, False, True]]` will return `ValueError: Item wrong length 3 instead of 824.`

In [25]:
# Therefore, booleans used with `[]` are typically generated by code, e.g.:
df[df["u_year"] == 2005]

u_company_name_id  u_year                       u_company_name  cb_naics  \
0                14651    2005         British American Tobacco PLC    312230   
15               13722    2005                 Bob Evans Farms Inc.    311612   
27               10161    2005                               BP PLC    324110   
42               10231    2005                         BT Group PLC    517311   
56               39131    2005                  Glaxosmithkline PLC    325412   
71               56071    2005  LVMH Moet Hennessy Louis Vuitton SE      3152   
86               36437    2005                          Freddie Mac    522298   
101               4304    2005                           Allianz SE    524126   
115              29114    2005                   EOG Resources Inc.      2111   
130               9724    2005                              BASF SE       325   
144              25503    2005                           Daimler AG    336111   
169              87672    2005                           Siemens AG    999977   
185              48172    2005            Infinity Energy Resources    211120   
203              96809    2005                    Tivity Health Inc    621999   
218              93646    2005                             TOTAL SA    324110   
241               9149    2005           AvalonBay Communities Inc.    531110   
265              56843    2005      Learning Tree International Inc    611420   
279               9423    2005                           Axa, Paris    524113   
293              25871    2005                 Dassault Systemes SA    511210   
307              29138    2005                     EPIQ Systems Inc    511210   
318              28946    2005                          EDAP TMS SA    334510   
332             103775    2005        Waddell & Reed Financial Inc.    523930   
347              11568    2005                             Bayer AG       325   
361              15114    2005                            Bunzl PLC    424130   
376              26755    2005                Deutsche Lufthansa AG    481111   
390              28868    2005                              E.ON SE      2211   
404              23192    2005                       Continental AG      3363   
418             103373    2005                        Volkswagen AG    336111   
432              82265    2005                               Rwe AG      2211   
446              55828    2005                     L'Air Liquide SA    325120   
460              84304    2005                               Sanofi    325412   
474              77954    2005                   Publicis Groupe SA    541810   
488               3307    2005                     Air France - KLM    481111   
501              89121    2005                               Sodexo    722310   
516             101235    2005                             Valeo SA    336390   
530              15715    2005                                  CGG    541360   
545              82492    2005                               SAP SE    511210   
560             102101    2005                         Vernalis PLC    325412   
571               7343    2005                     Arm Holdings PLC    533110   
582              62726    2005                    MicroStrategy Inc    511210   
597              94839    2005                       Technicolor SA    334220   
611              48149    2005             Infineon Technologies AG    334413   
626              66396    2005      Natural Gas Services Group Inc.    532412   
640               7783    2005        Ashford Hospitality Trust Inc    531190   
655              27790    2005               Dolby Laboratories Inc    533110   
699              66156    2005                    National Grid plc      2211   
713              36527    2005   Fresenius Medical Care AG &Co KGAA    621492   
727               3417    2005                           Aixtron SE    333242   
750              69671    2005      

In [26]:
# Let's take this apart, first, we generate a selection mask:
mask = df["u_year"] == 2005
# then we use it:
df[mask].head()

u_company_name_id  u_year                u_company_name  cb_naics u_iso3  \
0               14651    2005  British American Tobacco PLC    312230    GBR   
15              13722    2005          Bob Evans Farms Inc.    311612    USA   
27              10161    2005                        BP PLC    324110    GBR   
42              10231    2005                  BT Group PLC    517311    GBR   
56              39131    2005           Glaxosmithkline PLC    325412    GBR   

         u_fye   cb_cusip       cb_at      cb_ni  cb_financial_industry  
0   2005-12-31  110448107   32737.984   2707.110                  False  
15  2006-04-30  096761101    1209.183     54.774                  False  
27  2005-12-31  055622104  206914.000  22341.000                  False  
42  2006-03-31  05577E101   42960.710   2690.697                  False  
56  2005-12-31  37733W105   46747.922   8059.453                  False

In [27]:
# You can also use more complex boolean expressions to select rows that fulfill certain conditions
df[(df["u_year"] == 2005) & (df["u_iso3"] == "DEU")].head()

u_company_name_id  u_year u_company_name  cb_naics u_iso3       u_fye  \
101               4304    2005     Allianz SE    524126    DEU  2005-12-31   
130               9724    2005        BASF SE       325    DEU  2005-12-31   
144              25503    2005     Daimler AG    336111    DEU  2005-12-31   
169              87672    2005     Siemens AG    999977    DEU  2005-09-30   
347              11568    2005       Bayer AG       325    DEU  2005-12-31   

      cb_cusip        cb_at     cb_ni  cb_financial_industry  
101  018820100  1181690.680  5186.796                   True  
130  055262505    42240.532  3560.534                  False  
144  D1668R123   238772.614  3376.154                  False  
169  826197501   103945.989  2710.638                  False  
347  072730302    43486.192  1891.167                  False

# Select rows and columns using `df.loc[]`

`.loc[]` can select either rows, or columns, or both at the same time.

## Selecting rows using `loc`

Just like with `[]`, we can pass a boolean mask to `loc[]` to select rows programmatically:

In [28]:
df.loc[df["u_year"] == 2005].head()

u_company_name_id  u_year                u_company_name  cb_naics u_iso3  \
0               14651    2005  British American Tobacco PLC    312230    GBR   
15              13722    2005          Bob Evans Farms Inc.    311612    USA   
27              10161    2005                        BP PLC    324110    GBR   
42              10231    2005                  BT Group PLC    517311    GBR   
56              39131    2005           Glaxosmithkline PLC    325412    GBR   

         u_fye   cb_cusip       cb_at      cb_ni  cb_financial_industry  
0   2005-12-31  110448107   32737.984   2707.110                  False  
15  2006-04-30  096761101    1209.183     54.774                  False  
27  2005-12-31  055622104  206914.000  22341.000                  False  
42  2006-03-31  05577E101   42960.710   2690.697                  False  
56  2005-12-31  37733W105   46747.922   8059.453                  False

## Select columns using `loc`

When using `loc[]`, we can select columns by their labels, but we must specify a row selection as well or a placeholder `:`

In [29]:
df.loc[:, ["u_year", "u_company_name"]]

u_year                u_company_name
0      2005  British American Tobacco PLC
1      2006  British American Tobacco PLC
2      2007  British American Tobacco PLC
3      2008  British American Tobacco PLC
4      2009  British American Tobacco PLC
..      ...                           ...
819    2017           Midatech Pharma PLC
820    2018           Midatech Pharma PLC
821    2016                 Motif Bio plc
822    2017                 Motif Bio plc
823    2018                 Motif Bio plc

[824 rows x 2 columns]

## Select rows and columns using `loc`

In [30]:
df.loc[df["u_year"] > 2007, ["u_company_name", "u_year"]]

u_company_name  u_year
3    British American Tobacco PLC    2008
4    British American Tobacco PLC    2009
5    British American Tobacco PLC    2010
6    British American Tobacco PLC    2011
7    British American Tobacco PLC    2012
..                            ...     ...
819           Midatech Pharma PLC    2017
820           Midatech Pharma PLC    2018
821                 Motif Bio plc    2016
822                 Motif Bio plc    2017
823                 Motif Bio plc    2018

[673 rows x 2 columns]

# Exercise

*Preparations*
* create new jupyter notebook
* store it under notebooks/own_exercises

*Tasks*
1. Load the first sheet of the Excel file "wdi_reduced.xlsx" into a pandas DataFrame (see [here](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html) for help with `pandas.read_excel()`). *Hint*: you need to save the notebook in a suitable directory before you can use relative file paths! Also, if you use a different folder from the one specified in *preparations*, you may have to adjust the relative path, e.g. the number of times you use the "../" prefix.
2. What are the data types of the columns?
3. Display the descriptive statistics of the numeric columns.
4. Select the columns *countryname* and *year* from the DataFrame.
5. Select the rows from Germany and France.
6. Bonus: Select the rows from Germany and France in the years 2008 to 2010 and all columns starting with 'country' as well as *year*.
7. Bonus: Select only numeric columns.
8. Bonus: Select the numeric columns except those starting with "NY_".


# Side-Note: using the ydata-profiling package

The company [YData](https://ydata.ai/) is providing a package called [ydata-profiling](https://github.com/ydataai/ydata-profiling) (formerly known as pandas-profiling) that can be used to generate a report about a DataFrame. This report includes information about data types, missing values, distributions of values, correlations, and more.

In the past, the package has been prone to peculiar dependencies. Therefore, I recommend you setup a separate virtual environment.

Open a terminal in the project directory, then run:

**Windows**
```powershell
uv venv --python=3.12 .venv-ydata
Set-ExecutionPolicy -Scope Process -ExecutionPolicy Bypass
.\.venv-ydata\Scripts\activate
uv pip install setuptools legacy-cgi htmlmin ydata-profiling ipykernel jupyter
```
**Linux / MacOS**
```bash
uv venv --python=3.12 .venv-ydata
source .venv-ydata/bin/activate
uv pip install setuptools legacy-cgi htmlmin ydata-profiling ipykernel jupyter
```
Then, in a jupyter notebook:

In [31]:
# Optional: only runs if ydata_profiling is installed
try:
    import pandas as pd
    from ydata_profiling import ProfileReport
    df = pd.read_csv("../../data/raw/financial_data_intro.csv")
    profile = ProfileReport(df, title="Pandas Profiling Report", explorative=True)
    profile
except ImportError:
    print("ydata_profiling not installed - skipping this optional example")

ydata_profiling not installed - skipping this optional example
